system/lib/wasmfs/js_api.cpp

```diff
@@ -137,6 +137,22 @@ int _wasmfs_unlink(char* path) {
   return __syscall_unlinkat(AT_FDCWD, (intptr_t)path, 0);
 }
 
+int _wasmfs_truncate(char *path, off_t length) {
+  int err = __syscall_truncate64((intptr_t)path, length);
+  if (err == -1) {
+    return errno;
+  }
```

Comment:
Syscalls don't set errno, so this doesn't look right. Syscalls return negative error codes and libc handles those by setting errno appropriately. I would expect this code to be checking whether `err` is a negative error code (not necessarily -1) instead.

Reply:
Is it possible that `errno` is being set by `libc` behind the scenes? I've been doing this for the other `js_api.cpp` functions and the error codes match with the legacy API.

Reply:
Ah, no, all those places should be fixed. For example, in `_wasmfs_mknod`, it checks whether `__syscall_mknodat` returns -1, and if so returns `errno`. But `__syscall_mknodat` never sets errno and doesn't return -1 on errors; instead, it either returns `-EINVAL` or `-EPERM` or other negative error codes that `doOpen` can return.